### PR TITLE
feat: couchbase integration

### LINKED ISSUE #5178: feat: support Couchbase as a vector store
### Self Checks

- [X] I have searched for existing issues [search for existing issues](https://github.com/langgenius/dify/issues), including closed ones.
- [X] I confirm that I am using English to submit this report (我已阅读并同意 [Language Policy](https://github.com/langgenius/dify/issues/1542)).
- [X] Please do not modify this template :) and fill in all the required fields.

### 1. Is this request related to a challenge you're experiencing?

No

### 2. Describe the feature you'd like to see

I want to add support for [Couchbase as a vector store](https://docs.couchbase.com/server/current/vector-search/vector-search.html). 

Couchbase is already supported in [LangChain](https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/vectorstores/couchbase.py) and [LangFlow](https://github.com/langflow-ai/langflow/blob/dev/src/backend/base/langflow/components/vectorsearch/CouchbaseSearch.py), among others.

The following configuration parameters will likely need to be added:
- `couchbase_connection_string`
- `couchbase_username`
- `couchbase_password`
- `bucket_name`
- `scope_name`
- `collection_name`
- `index_name

### 3. How will this feature improve your workflow or experience?

So that myself and other users can opt to use Couchbase as their vector store in Dify.

### 4. Additional context or comments

Update documentation to include Couchbase, if needed

### 5. Can you help us with this feature?

- [X] I am interested in contributing to this feature.

### PR DESCRIPTION
**Pull request by roadgoat19**
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [x] Please open an issue before creating a PR or link to an existing issue
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue. Close issue syntax: `Fixes #<issue number>`, see [documentation](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more details.

Fixes #5178

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- Tested test_vdb, docker-compose.yaml, docker-compose.middleware.yaml

### CODE REVIEW

**Comment by roadgoat19** [comment]
Hello @JohnJyong, do you have any comments? The only issue left to be fixed is disabling embedding models with a dimensionality > 2048. I'd like to ideally throw an error like this but I'm not sure how other vdb's like relyt are handling it 
Screenshot 2024-07-16 at 11 49 20 AM

**Comment by crazywoola** [comment]
@roadgoat19 Hello , sorry for the late response, we have working on some new features regarding RAG recently. Please resolve the conflicts and we can proceed with this PR. :)

**Comment by roadgoat19** [comment]
@crazywoola I've fixed a lingering merge conflict, could you rerun the workflows please?

**Comment by crazywoola** [comment]
Hello the API Tests (3.10) failed and there are some lint errors in the superlinter. 
You could run `./dev/reformat` to resolve linting issues.

**Comment by roadgoat19** [comment]
@crazywoola Hello I believe I've fixed this issue causing the api tests to fail. However, I am a little unsure about how to fix the linter errors in the .env.example. When I run dev/reformat, it says that basically every single key is out of order. This issue persists even with the .env.example taken straight from the current main branch of dify. Is there a command I can run to fix this issue /or how should I proceed?

**Review by crazywoola — DISMISSED** [review]

**Comment by crazywoola** [comment]
Hello there are some conflicts here. And still got a lint error. I think you need to upgrade ruff and sync deps by `./dev/sync-poetry`.

**Comment by ejscribner** [comment]
Thanks @crazywoola, and sorry for the delay here. I'm picking this PR up for @roadgoat19, and have resolved all conflicts and updated the dependencies. Let me know if anything else needs adjusting!

**Comment by crazywoola** [comment]
Hello @ejscribne Thanks for picking this up, please resolve those conflicts again :)

**Comment by ejscribner** [comment]
Hey @crazywoola just fixed the conflicts, let me know if anything else needs to be done. Thanks!

**Comment by crazywoola** [comment]
I have resolved the conflicts in poetry.lock and fixed lint errors as well.

File: api/pyproject.toml [hunk 1]
```diff
@@ -198,7 +198,7 @@ safetensors = "~0.4.3"
 ############################################################
 # [ Tools ] dependency group
 ############################################################
-[tool.poetry.group.tools.dependencies]
+[tool.poetry.group.tool.dependencies]
```
Comment (bowenliang123) [inline api/pyproject.toml]:
```suggestion
[tool.poetry.group.tools.dependencies]
```
Revert the change in depenceny group first. cc @crazywoola

File: api/pyproject.toml [hunk 2]
```diff
@@ -236,7 +236,8 @@ alibabacloud_gpdb20160503 = "~3.8.0"
 alibabacloud_tea_openapi = "~0.3.9"
 chromadb = "0.5.1"
 clickhouse-connect = "~0.7.16"
-elasticsearch = "8.14.0"
+couchbase = "^4.3.0"
```
Comment (bowenliang123) [inline api/pyproject.toml]:
```suggestion
couchbase = "~4.3.0"
```
The `^` operator is too wide and forbidden.

File: docker/.env.example [hunk 3]
```diff
@@ -345,8 +345,8 @@ VOLCENGINE_TOS_REGION=your-region
 # Vector Database Configuration
 # ------------------------------
 
-# The type of vector store to use. 
-# Supported values are `weaviate`, `qdrant`, `milvus`, `myscale`, `relyt`, `pgvector`, `pgvecto-rs`, ``chroma`, `opensearch`, `tidb_vector`, `oracle`, `tencent`, `elasticsearch`, `analyticdb`.
+# The type of vector store to use.
+# Supported values are `weaviate`, `qdrant`, `milvus`, `myscale`, `relyt`, `pgvector`, `pgvecto-rs`, ``chroma`, `opensearch`, `tidb_vector`, `oracle`, `tencent`, `elasticsearch`, `analyticdb`, `couchbase`, `couchbase`.
```
Comment (bowenliang123) [inline docker/.env.example]:
```suggestion
# Supported values are `weaviate`, `qdrant`, `milvus`, `myscale`, `relyt`, `pgvector`, `pgvecto-rs`, ``chroma`, `opensearch`, `tidb_vector`, `oracle`, `tencent`, `elasticsearch`, `analyticdb`, `couchbase`.
```
Why repeating `couchbase` here?

**Comment by crazywoola** [comment]
@ejscribner https://github.com/langgenius/dify/actions/runs/11285975856/job/31389592906?pr=6165 The tests are failed.

**Comment by ejscribner** [comment]
Hey @crazywoola sorry for the delay fixing this, added a healthcheck to be sure the resource is ready for the tests. It should fix the issue.

**Review by crazywoola — APPROVED** [review]

**Comment by crazywoola** [comment]
Finally
